pkg/webhook/mutating/server.go

```diff
@@ -0,0 +1,102 @@
+	"github.com/pkg/errors"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const (
+	webhookName      = "nmstate"
+	RetrievalTimeout = 120
+)
+
+type server struct {
+	mgr           manager.Manager
+	webhookServer *webhook.Server
+	log           logr.Logger
+}
+
+// Add creates a new Conditions Mutating Webhook and adds it to the Manager. The Manager will set fields on the Webhook
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newServer(mgr))
+}
+
+func newServer(mgr manager.Manager) *server {
+	s := &server{
+		webhookServer: &webhook.Server{
+			Port:    8443,
+			CertDir: "/etc/webhook/certs/",
+		},
+		mgr: mgr,
+		log: logf.Log.WithName("webhook/mutating/server"),
+	}
+	s.webhookServer.Register("/mutating", resetConditionsHook())
+	return s
+}
+
+// add adds a new Webhook to mgr with r as the webhook.Server
+func add(mgr manager.Manager, s *server) error {
+	mgr.Add(s)
+	return nil
+}
+
+func (s *server) updateCABundle() error {
+	s.log.Info("Updating CA bundle for webhook")
+	mutatingWebHook := admissionregistrationv1beta1.MutatingWebhookConfiguration{}
+
+	// Do some polling to wait for manifest to be deployed
+	err := wait.PollImmediate(1*time.Second, RetrievalTimeout, func() (bool, error) {
+		webhookKey := types.NamespacedName{Name: webhookName}
+		err := s.mgr.GetClient().Get(context.TODO(), webhookKey, &mutatingWebHook)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed retrieving mutationg webhook "+webhookName)
+	}
+	// Retrieve cluster CA bundle and encode to base 64
+	authenticationConfig := corev1.ConfigMap{}
+	err = s.mgr.GetClient().Get(context.TODO(), types.NamespacedName{Namespace: "kube-system", Name: "extension-apiserver-authentication"}, &authenticationConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve cluster authentication config")
+	}
+	clientCaFile := authenticationConfig.Data["client-ca-file"]
+	// Update the CA bundle at webhook
+	mutatingWebHook.Webhooks[0].ClientConfig.CABundle = []byte(clientCaFile)
+	err = s.mgr.GetClient().Update(context.TODO(), &mutatingWebHook)
+	if err != nil {
+		return errors.Wrap(err, "failed to update mutating webhook CABundle")
+	}
+	return nil
+}
+
+func (s *server) Start(stop <-chan struct{}) error {
+	err := s.updateCABundle()
+	if err != nil {
+		return errors.Wrap(err, "failed updating CA bundle at webhook")
+	}
+	return s.webhookServer.Start(stop)
+}
+
+func (s *server) InjectFunc(f inject.Func) error {
+	return s.webhookServer.InjectFunc(f)
+}
+
+func (s *server) NeedLeaderElection() bool {
```

Comment:
What does this do? We don't need leader election

Reply:
Since we are delegating on webhook.Server we have to honor https://github.com/kubernetes-sigs/controller-runtime/blob/40070e2a1958c3d974ba95da883a2bd088137789/pkg/manager/internal.go#L215